File: packages/sdks/flutter/lib/data/contracts/http_client_contract.dart
import 'package:sdk_flutter/core/either/either.dart';
import 'package:sdk_flutter/core/types/json.dart';

abstract class HttpClientContract {
  FutureEither<bool> delete({required String url, Json? urlParams});

  FutureEither<T> get<T>({
    required String url,
    Json? queryParams,
    Json? urlParams,
  });

  FutureEither<T> index<T>({
    required String url,
    Json? queryParams,
    Json? urlParams,
  });

  FutureEither<bool> patch({
    required String url,
    required Json data,
    Json? urlParams,
  });

  FutureEither<Json> post({
    required String url,
    required dynamic data,
    Json? urlParams,
  });

  FutureEither<bool> put({
    required String url,
    required dynamic data,
    Json? urlParams,
  });
}

File: packages/sdks/react/data/contracts/http_client_contract.ts
import { EitherPromise } from "../../core/either/either";

export type Json = Record<string, any>;

export class PaginatedData<T> {
    data: T[];
    meta: Meta;

    constructor(data: T[], meta: Meta) {
        this.data = data;
        this.meta = meta;
    }

    static fromMap<T>(data: Json): PaginatedData<T> {
        return new PaginatedData(data.data, data.meta);
    }
}

export type Meta = {
    currentPage: number;
    lastPage: number;
    perPage: number;
    total: number;
}

export abstract class HttpClientContract {
    delete: (url: string) => EitherPromise<boolean>;

    get: (url: string, queryParams?: Json) => EitherPromise<Json>;

    index: (url: string, queryParams?: Json, isPaginated?: boolean) => EitherPromise<Json[] | PaginatedData<Json[]>>;

    patch: (url: string, data: Json) => EitherPromise<boolean>;

    post: (url: string, data: Json) => EitherPromise<Json>;

    put: (url: string, data: Json) => EitherPromise<boolean>;
}
